streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java

```diff
@@ -146,14 +132,23 @@ private void flush(final LRUNode evicted) {
                 deleted.add(node.key);
             }
         }
-        // clear dirtyKeys before the listener is applied as it may be re-entrant.
+
         dirtyKeys.clear();
-        listener.apply(entries);
+        applyFlushListener(entries);
         for (final Bytes key : deleted) {
             delete(key);
         }
     }
 
+    // Entries should be removed from dirtyKeys before the listener is applied as it may be re-entrant
+    private void applyFlushListener(final List<ThreadCache.DirtyEntry> entries) {
```

Comment:
Basically just refactored things so the actual forwarding downstream is done only on the passed in dirty entries (which is all that needs to happen for an evicted entry), and the actual `flush` method is responsible for, well, flushing the entire cache